source/Halibut/Transport/Streams/TimeoutStreamWrapper.cs

```diff
@@ -0,0 +1,192 @@
+                nameof(WriteAsync),
+                cancellationToken).ConfigureAwait(false);
+        }
+
+        public override async Task<int> ReadAsync(byte[] buffer, int offset, int count, CancellationToken cancellationToken)
+        {
+            return await WrapWithCancellationAndTimeout(
+                async ct => await inner.ReadAsync(buffer, offset, count, ct).ConfigureAwait(false),
+                CanTimeout ? ReadTimeout : int.MaxValue,
+                nameof(ReadAsync),
+                cancellationToken).ConfigureAwait(false);
+        }
+
+        public override async Task WriteAsync(byte[] buffer, int offset, int count, CancellationToken cancellationToken)
+        {
+            await WrapWithCancellationAndTimeout(
+                async ct =>
+                {
+                    await inner.WriteAsync(buffer, offset, count, ct).ConfigureAwait(false);
+                    return 0;
+                },
+                CanTimeout ? WriteTimeout : int.MaxValue,
+                nameof(WriteAsync),
+                cancellationToken).ConfigureAwait(false);
+        }
+
+#if !NETFRAMEWORK
+        public override async ValueTask<int> ReadAsync(Memory<byte> buffer, CancellationToken cancellationToken = default)
+        {
+            return await WrapWithCancellationAndTimeout(
+                async ct => await inner.ReadAsync(buffer, ct).ConfigureAwait(false),
+                CanTimeout ? ReadTimeout : int.MaxValue,
+                nameof(ReadAsync),
+                cancellationToken).ConfigureAwait(false);
+        }
+
+        public override async ValueTask WriteAsync(ReadOnlyMemory<byte> buffer, CancellationToken cancellationToken = default)
+        {
+            await WrapWithCancellationAndTimeout(
+                async ct =>
+                {
+                    await inner.WriteAsync(buffer, ct).ConfigureAwait(false);
+                    return 0;
+                },
+                CanTimeout ? WriteTimeout : int.MaxValue,
+                nameof(WriteAsync),
+                cancellationToken).ConfigureAwait(false);
+        }
+#endif
+
+        async Task<T> WrapWithCancellationAndTimeout<T>(Func<CancellationToken, Task<T>> action, int timeout, string methodName, CancellationToken cancellationToken)
+        {
+            using var timeoutCancellationTokenSource = new CancellationTokenSource(timeout);
+            using var linkedCancellationTokenSource = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken, timeoutCancellationTokenSource.Token);
+
+            var actionTask = action(linkedCancellationTokenSource.Token);
+            var cancellationTask = linkedCancellationTokenSource.Token.AsTask<T>();
+
+            try
+            {
+                var completedTask = await Task.WhenAny(actionTask, cancellationTask).ConfigureAwait(false);
+
+                if (completedTask == cancellationTask)
+                {
+                    actionTask.IgnoreUnobservedExceptions();
```

Comment:
Is there a way we can get `actionTask` to die if we get here? I can't think of a nice way.

Maybe if we create another task using Task.Run, and pass in the cancellation token to that. Would that kill the process running the actual thread from `action`?

Reply:
No you can't make actionTask stop apart from making sure it cancels co-operatively!